### DISCUSSION TITLE
Csla 6 PrivateField getter User.Identity is null on saving of object

### DISCUSSION THREAD

**Discussion by JacoJordaan** [MarimerLLC/csla#3213] [Questions]
When I call save on my business object, the execution of the Get of the private field property is called. In my case, I am creating a new object for assignment to the backing field. The creation of the new object fails with an authorization exception. This is caused by the ApplicationContext.User.Identity value that is null.

Is there a way to prevent csla to call the getter of a private field on saving of the object?

````
#pragma warning disable CSLA0007
        public static readonly PropertyInfo<CustomerPOCustomerList> CustomerListProperty = RegisterProperty<CustomerPOCustomerList>(c => c.CustomerList, RelationshipTypes.PrivateField);

        [NotUndoable, NonSerialized]
        private CustomerPOCustomerList _customerList = CustomerListProperty.DefaultValue;

        public CustomerPOCustomerList CustomerList
        {
            get
            {
                if (_customerList == null)
                {
                    var dataPortalFactory = ApplicationContext.GetRequiredService<IDataPortalFactory>();
                    //Calling the following throws an authorization exception in Csla 6 - the ApplicationContext.User.Identity is null
                    _customerList = new CustomerPOCustomerList.Factory(dataPortalFactory).GetCustomerPOCustomerList(CustomerId);

                }
                return GetProperty(CustomerListProperty, _customerList);
            }
            private set
            {
                SetProperty(CustomerListProperty, ref _customerList, value);
            }
        }

#pragma warning restore CSLA0007
````

**Comment by rockfordlhotka** [comment]
This is .NET 6 with a remote data portal?

Do you have a rule attached to this property (or one that uses the property)? Rules are now re-run as the object graph arrives on the app server, and that could be what's reading the property.

Another possibility is that something like `DataMapper` could be reading the property in your data portal operation method.

**Reply by JacoJordaan** [reply]
I have tested this with the Blazor example and the `LocalProxy.cs` and `GraphMerger.cs` is calling the property get. 
It does not check for the private field.

In `LocalProxy` method `private void SetApplicationContext(IUseApplicationContext useApplicationContext, ApplicationContext applicationContext)`:
````
...
 else if (typeof(IEnumerable).IsAssignableFrom(managedProperty.Type))
              {
                //property is a list and needs to be processed (could be a Csla.Core.MobileList or something like that)
                var enumerable = (IEnumerable)target.ReadProperty(managedProperty);
                if (enumerable != null)
                {
                  foreach (var item in enumerable)
                  {
                    SetApplicationContext(item, applicationContext);
                  }
                }
              }
...
````
In `GraphMerger.cs` method `public void MergeGraph(IEditableBusinessObject target, IEditableBusinessObject source)`
````
...
        var targetProperties = imp.GetManagedProperties();
        foreach (var item in targetProperties)
        {
          var sourceFieldExists = true;
          if (sourceFieldManager != null && (item.RelationshipType & RelationshipTypes.LazyLoad) == RelationshipTypes.LazyLoad)
            sourceFieldExists = sourceFieldManager.FieldExists(item);
          object sourceValue = null;
          if (sourceFieldExists)
            sourceValue = ReadProperty(source, item);
...
````

**Comment by JacoJordaan** [comment]
I am calling this from a MSTest project, so it will be a local data portal. All .Net6 projects.

When calling the same test in Csla5, it is not calling the property get.

I have commented out all business rules and authorization rules.

What is interesting is that for a simple data type e.g. int, DateTime, etc it is not calling the property get, but for any complex data type e.g. List<int>, csla object, it is calling the property get.

Here are examples:
NOT calling the property get on save:
````
        public static readonly PropertyInfo<int> Test1Property = RegisterProperty<int>(c => c.Test1, RelationshipTypes.PrivateField);

        [NotUndoable, NonSerialized]
        private int _test1 = Test1Property.DefaultValue;

        public int Test1
        {
            get
            {
                return GetProperty(Test1Property, _test1);
            }
            private set
            {
                OnPropertyChanged(Test1Property);
            }
        }
````

Calling of property get on save:
````
        public static readonly PropertyInfo<List<int>> Test2Property = RegisterProperty<List<int>>(c => c.Test2, RelationshipTypes.PrivateField);

        [NotUndoable, NonSerialized]
        private List<int> _test2 = Test2Property.DefaultValue;

        public List<int> Test2
        {
            get
            {
                return GetProperty(Test2Property, _test2);
            }
            private set
            {
                OnPropertyChanged(Test2Property);
            }
        }
````

**Reply by rockfordlhotka** [reply]
Narrowing it to `GrapherMerger` is extremely helpful, thank you!

I haven't had time to look deeper, but knowing where to look will be extremely helpful.

It is possible that, because your property references what amounts to a child object, and the `PropertyInfo` isn't decorated with the child flag (only the private field flag), `GraphMerger` isn't handling the property correctly? I don't know, just guessing at this point.

**Reply by JacoJordaan** [reply]
Let me try to explain my use case.
I want to lazy load an object that will set a complex property value only when I call the get of the property. 
I don't want to serialize the property, so I am using the private field enum. By default csla don't serialize private fields, so this works well. The property will most of the time only be needed on the client, but if required it can be reloaded on the server via the get e.g. if used in a rule in some scenarios. 

There is not a way to set the property as both lazy load and private field. 

For the above reasons I don't want the get to be called in localproxy or graphmerger. 
Currently I don't use the graphmerger in csla5 when saving the object in some scenarios. Mostly for performance reasons. 
It will be helpful to use this it csla6.
In csla6 it breaks in local proxy and graph merger. 

I have not tested this, but it should probably work to do a check in both graphmerger and localproxy to see if the source value is null if this is a private field. If null then dont call the get functionality since it is not required.

**Reply by JacoJordaan** [reply]
For some reason, I thought it is not possible to have both PrivateField and LazyLoad decoration on a property.
I have added the LazyLoad decoration to the property and now it is working fine. It helps to look at the actual code.
